include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -330,6 +346,32 @@ MatrixFree<dim, Number, VectorizedArrayType>::copy_from(
 
 
 
+namespace internal
+{
+  template <typename Number,
+            typename Number2,
+            typename std::enable_if<std::is_same<Number, Number2>::value,
+                                    bool>::type = true>
+  const dealii::AffineConstraints<Number> *
+  store_affine_constraints(
+    const dealii::AffineConstraints<Number2> *affine_constraints)
+  {
+    return affine_constraints;
+  }
+
+  template <typename Number,
+            typename Number2,
+            typename std::enable_if<!std::is_same<Number, Number2>::value,
+                                    bool>::type = true>
+  const dealii::AffineConstraints<Number> *
+  store_affine_constraints(const dealii::AffineConstraints<Number2> *)
+  {
+    return nullptr;
+  }
```

Comment:
Do we really need `std::enable_if` here? Wouldn't it be enough to write the function where we want to do something as:
```c++
  template <typename Number>
  const dealii::AffineConstraints<Number> *
  store_affine_constraints(
    const dealii::AffineConstraints<Number> *affine_constraints)
  {
    return affine_constraints;
  }
```
because the compiler needs to choose to more specialized version of the two?

Reply:
Good question, I will investigate this!

Reply:
Do you mean something like this?
```cpp
  template <typename Number, typename Number2>
  const dealii::AffineConstraints<Number> *
  store_affine_constraints(
    const dealii::AffineConstraints<Number2> *affine_constraints)
  {
    return affine_constraints;
  }

  template<>
  const dealii::AffineConstraints<double> *
  store_affine_constraints(
    const dealii::AffineConstraints<float> *)
  {
    return nullptr;
  }

  template<>
  const dealii::AffineConstraints<float> *
  store_affine_constraints(
    const dealii::AffineConstraints<double> *)
  {
    return nullptr;
  }
```

Reply:
Or like this?
```cpp
  template <typename Number, typename Number2>
  const dealii::AffineConstraints<Number> *
  store_affine_constraints(
    const dealii::AffineConstraints<Number2> *)
  {
    return nullptr;
  }

  template<typename Number>
  const dealii::AffineConstraints<Number> *
  store_affine_constraints(
    const dealii::AffineConstraints<Number> *affine_constraints)
  {
    return affine_constraints;
  }
```

Reply:
I was thinking of the second variant, if it is possible in terms of the compiler selecting the right interface. Alternatively, you could put the **store** part into the function as well, handing in the other pointer, which allows you to automatically deduce the template arguments at the call site and avoid the possible ambiguity.

Reply:
Apparently partial function template specialization is not allowed by the standard. I propose to keep the `std::enable_if_t` solution. The first version with full (explicit) template specialization would also work.

Reply:
I am not against `enable_if` in general, but we should only use it if necessary. Here it is not in my opinion. Have you tried this code:
```c++
namespace internal
{
  template <typename Number, typename Number2>
  void
  store_affine_constraints(
    const dealii::AffineConstraints<Number2> *,
    SmartPointer<const dealii::AffineConstraints<Number>> &stored_constraints)
  {
    stored_constraints = nullptr;
  }

  template <typename Number>
  void
  store_affine_constraints(
    const dealii::AffineConstraints<Number> *              affine_constraints,
    SmartPointer<const dealii::AffineConstraints<Number>> &stored_constraints)
  {
    stored_constraints = affine_constraints;
  }
} // namespace internal
```
and then call it like this:
```
  internal::store_affine_constraints(constraints[no], affine_constraints[no]);
```
As an additional advantage besides obviating the use of `enable_if`, you do not need to specify the function template arguments, which automatically deduces the arguments. I also find this implementation in the helper function more intuitive when the name is `store_affine_constraints`. In your code, I would probably lean towards using `extract_affine_constraints_as_number_pointer`, because the function itself does not store anything.

Reply:
This works on my machine but I am not sure if every compiler is able to produce the same result here. From what I have read this is something you should not do since this is not supported by the standard, for example (https://stackoverflow.com/questions/8061456/c-function-template-partial-specialization?noredirect=1&lq=1). Let's see what the CI says.

Reply:
What I suggest is function overloading, not partial template specialization, which follows clear rules in terms of which of the two versions should get selected.

Reply:
I see... what a beautiful programming language 😄